src/Build/Graph/ProjectGraph.cs

```diff
@@ -54,5 +59,139 @@ public ProjectGraph(string entryProjectFile)
         /// Get an unordered collection of all project nodes in the graph.
         /// </summary>
         public IReadOnlyCollection<ProjectGraphNode> ProjectNodes => _projectNodes;
+
+        /// <summary>
+        /// Gets the target list for every node, given a particular target list for the entry project.
+        /// </summary>
+        /// <param name="entryProjectTargets">The target list for the entry project. May be null or empty, in which case the project's default targets will be used.</param>
+        /// <returns>A dictionary containing the target list for each node.</returns>
+        public IReadOnlyDictionary<ProjectGraphNode, string[]> GetTargetLists(string[] entryProjectTargets)
+        {
+            // Seed the dictionary with empty lists for every node.
+            Dictionary<ProjectGraphNode, List<string>> targetLists = _projectNodes.ToDictionary(node => node, node => new List<string>());
```

Comment:
Worth using a singleton for empty lists?

Reply:
Nvm, it gets mutated later on :)

---
In reply to: [218880779](https://github.com/Microsoft/msbuild/pull/3769#discussion_r218880779) [](ancestors = 218880779)